src/lib/nullifier.ts

```diff
@@ -0,0 +1,84 @@
+import {
+  Poseidon,
+  PublicKey,
+  Field,
+  Group,
+  Struct,
+  MerkleMapWitness,
+} from 'snarkyjs';
+import { Nullifier as JsonNullifier } from '../mina-signer/src/TSTypes.js';
+
+export { Nullifier };
+
+/*
+RFC: https://github.com/o1-labs/snarkyjs/issues/756
+Paper: PLUME https://eprint.iacr.org/2022/1255.pdf
+*/
+class Nullifier extends Struct({
+  message: Field,
+  publicKey: PublicKey,
+  public: {
+    nullifier: Group,
+    s: Field,
+  },
+  private: {
+    c: Field,
+    g_r: Group,
+    h_m_pk_r: Group,
+  },
+}) {
+  static fromJSON(json: JsonNullifier): Nullifier {
+    return super.fromJSON(json as any) as Nullifier;
+  }
+
+  /**
+   * Verifies the correctness of the Nullifier. Throws an error if the Nullifier is incorrect.
+   */
+  verify() {
+    let {
+      message,
+      publicKey,
+      public: { nullifier },
+      private: { g_r, h_m_pk_r, c },
+    } = this;
+
+    let G = Group.generator;
+
+    let {
+      x,
+      y: { x0 },
+    } = Poseidon.hashToGroup([message, ...publicKey.toFields()]);
+
+    Poseidon.hash([
+      ...Group.toFields(G),
+      ...publicKey.toFields(),
+      x,
+      x0,
+      ...Group.toFields(nullifier),
+      ...Group.toFields(g_r),
+      ...Group.toFields(h_m_pk_r),
```

Comment:
The way you use `g_r` and `h_m_pk_r` here makes the check not sufficient: You don't prove that there's a connection between `g_r`, `h_m_pk_r` on the one hand and `pk`, `c`, `nullifier` and `s` on the other hand.

So, at the moment, I can create a **valid nullifier without owning the private key**, by doing the following:
* create any random group elements `g_r` and `h_m_pk_r`
* computing `hash[m, pk] = x, x0` from the message and public key, which I have access to
* recreate the `c` hash
* put all of those together into a `Nullifier`

So, it's not secure yet!

I know that the paper says that `g_r` and `h_m_pk_r` are
> ...included here to increase prover efficiency via precomputation

I'm not sure how using them to increase efficiency without losing security works, yet.
But I would ask that we first try to make it work without those values, and re-compute everything, like `g^s / pk^c` etc. We can worry about introducing efficiency tricks later

Reply:
Ooof that part confused me a bit, I thought by "optimizing the computation" it meant we could just take these values directly :D good catch